gemini/sample-apps/llamaindex-rag/common/utils.py

```diff
@@ -0,0 +1,248 @@
+# Copyright 2024 Google, LLC. This software is provided as-is, without
+# warranty or representation for any use or purpose. Your use of it is
+# subject to your agreement with Google.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#    http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+GCP Download utilities
+"""
+import logging
+import os
+
+from google.cloud import storage
+from llama_index.core import Document
+from llama_index.core.schema import NodeRelationship, RelatedNodeInfo, TextNode
+import yaml
+
+logging.basicConfig(level=logging.INFO)  # Set the desired logging level
+logger = logging.getLogger(__name__)
+
+
+class Blob:
+    def __init__(self, path: str, mimetype: str):
+        self.path = path
+        self.mimetype = mimetype
+
+
+def download_blob(bucket_name, source_blob_name, destination_file_name):
+    """Downloads a blob from the bucket."""
+    # The ID of your GCS bucket
+    # bucket_name = "your-bucket-name"
+
+    # The ID of your GCS object
+    # source_blob_name = "storage-object-name"
+
+    # The path to which the file should be downloaded
+    # destination_file_name = "local/path/to/file"
+
+    storage_client = storage.Client()
+
+    bucket = storage_client.bucket(bucket_name)
+
+    # Construct a client side representation of a blob.
+    # Note `Bucket.blob` differs from `Bucket.get_blob` as it doesn't retrieve
+    # any content from Google Cloud Storage. As we don't need additional data,
+    # using `Bucket.blob` is preferred here.
+    blob = bucket.blob(source_blob_name)
+    blob.download_to_filename(destination_file_name)
+
+    print(
+        "Downloaded storage object {} from bucket {} to local file {}.".format(
+            source_blob_name, bucket_name, destination_file_name
+        )
+    )
+
+
+def download_bucket_with_transfer_manager(
```

Comment:
Should the user be downloading all of these files to their local machine? Is there some way to load them remotely? Direct from GCS to LlamaIndex or wherever it's being used?

Reply:
This is only used for indexing the data initially. User run the `src/indexing/run_parse_embed_index.py` script to download files from the specified bucket in `common/config.yaml` and index into Vector Search